system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
     return new CLibCompiledXpath(xpath);
 }
 
+xmlNsPtr createNsFromXmlns(xmlNodePtr node, const char *attname, const char *uri, xmlNsPtr ns)
+{
+    if (*attname=='@')
+        attname++;
+    if (strncmp(attname, "xmlns", 5)==0)
+    {
+        attname+=5;
+        if (*attname==':')
+            attname++;
+    }
+    if (ns && streq((const char *)ns->href, uri)) //preprocessed the current prefix
+        return nullptr;
+    return xmlNewNs(node, (const xmlChar *)uri, (const xmlChar *)attname);
+}
+void addChildFromPtree(xmlNodePtr parent, IPropertyTree &tree, const char *name)
+{
+    if (isEmptyString(name))
+        name = tree.queryName();
+    if (*name==':') //invalid xml but maybe valid ptree?
+        name++;
+    StringAttr prefix;
+    const char *colon = strchr(name, ':');
+    if (colon)
+    {
+        prefix.set(name, colon-name);
+        name = colon+1;
+    }
+    xmlNodePtr node = xmlNewChild(parent, nullptr, (const xmlChar *)name, (const xmlChar *)tree.queryProp(nullptr));
+    if (!node)
+        return;
+    xmlNsPtr ns = nullptr;
+    if (tree.getAttributeCount())
+    {
+        Owned<IAttributeIterator> atts = tree.getAttributes();
+        ForEach(*atts)
+        {
+            const char *attname = atts->queryName()+1;
+            if (strncmp(attname, "xmlns", 5)==0)
+            {
+                attname+=5;
+                if (*attname==':')
+                    attname++;
+             xmlNewNs(node, (const xmlChar *)atts->queryValue(), (const xmlChar *)attname);
+            }
+            else
+                xmlNewProp(node, (const xmlChar *)attname, (const xmlChar *)atts->queryValue());
+        }
+    }
+    if (prefix.isEmpty())
+        node->ns = nullptr;
+    else
+    {
+        xmlNsPtr ns = xmlSearchNs(node->doc, node, (const xmlChar *)prefix.str());
+        if (!ns) //invalid but let's repair it
+            ns = xmlNewNs(node, (const xmlChar *)"urn:hpcc:unknown", (const xmlChar *)prefix.str());
+        node->ns = ns;
+    }
+    Owned<IPropertyTreeIterator> children = tree.getElements("*");
+    ForEach(*children)
+        addChildFromPtree(node, children->query(), nullptr);
+}
+
+IPropertyTree *createPtreeFromXmlNode(xmlNodePtr node);
+
+void copyXmlNode(IPropertyTree *tree, xmlNodePtr node)
+{
+    for(xmlAttrPtr att=node->properties;att!=nullptr; att=att->next)
+        tree->setProp((const char *)att->name, (const char *)xmlGetProp(node, att->name));
```

Comment:
Do we need to manage the memory of the xmlChar* returned from xmlGetProp here? Or is it just returning a pointer to the xmlChar* it retains ownership of?

Reply:
We own the block returned from xmlGetProp, and are required to free it.